### PR TITLE
NX adaptations

### PR DESCRIPTION
**Pull request by YunHsiao**
https://github.com/cocos-creator/3d-tasks/issues/10057

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -176,6 +176,9 @@ CMakeSettings.json
 cocos/platform/ohos/libcocos/build/.transforms
 cocos/platform/ohos/libcocos/build
 
+# NVN backend
+cocos/renderer/gfx-nvn
```
Comment (pandamicro) [inline .gitignore:180]: Use NX instead of NVN, and `gfx-nx` instead of `gfx-nvn`

File: CMakeLists.txt [hunk 2]
```diff
@@ -29,7 +29,9 @@ cc_set_if_undefined(USE_PHYSICS_PHYSX        OFF)
 cc_set_if_undefined(USE_MODULES              OFF)
 
 
-if(ANDROID OR WINDOWS)
+if(NINTENDO)
```
Comment (pandamicro) [inline CMakeLists.txt:32]: Use NX as platform identifier

File: CMakeLists.txt [hunk 3]
```diff
@@ -29,7 +29,9 @@ cc_set_if_undefined(USE_PHYSICS_PHYSX        OFF)
...
+    cc_set_if_undefined(CC_USE_NVN ON)
```
Comment (pandamicro) [inline CMakeLists.txt:33]: `CC_USE_NX` or `CC_USE_NINTENDOSDK`

File: CMakeLists.txt [hunk 4]
```diff
@@ -1120,6 +1117,48 @@ if(CC_USE_VULKAN)
...
+                    cocos/renderer/gfx-nvn/NVNPipelineState.h
+                    cocos/renderer/gfx-nvn/NVNShader.h
+                    cocos/renderer/gfx-nvn/NVNDescriptorSet.cpp
+                    cocos/renderer/gfx-nvn/NVNGPUContext.cpp
+                    cocos/renderer/gfx-nvn/NVNQueryPool.cpp
+                    cocos/renderer/gfx-nvn/NVNStd.cpp
+                    cocos/renderer/gfx-nvn/NVNDescriptorSet.h
+                    cocos/renderer/gfx-nvn/NVNGPUObjects.h
+                    cocos/renderer/gfx-nvn/NVNQueryPool.h
+                    cocos/renderer/gfx-nvn/NVNStd.h
```
Comment (pandamicro) [inline CMakeLists.txt:1158]: NX prefix

File: CMakeLists.txt [hunk 5]
```diff
@@ -2029,6 +2085,52 @@ target_include_directories(cocos2d PUBLIC
     ${CWD}/external/sources/khronos
 )
 
+if(NINTENDO)
+    target_compile_definitions(cocos2d PUBLIC CC_USE_NVN)
+
+    if(${CMAKE_SIZEOF_VOID_P} STREQUAL "4")
+        if(${MSVC_VERSION} LESS 1920)
+            set(NN_SDK_PLATFORM_FOLDER win32-v141)
```
Comment (pandamicro) [inline CMakeLists.txt:2093]: What does NN mean?

File: CMakeLists.txt [hunk 6]
```diff
@@ -2029,6 +2085,52 @@ target_include_directories(cocos2d PUBLIC
...
+      IMPORTED_LOCATION_RELEASE $ENV{NINTENDO_SDK_ROOT}/Libraries/${NN_SDK_PLATFORM_FOLDER}/Release/libnnSdk.lib
+    )
+
+    add_library(NN_SDK_DIAG STATIC IMPORTED GLOBAL)
+    set_target_properties(NN_SDK_DIAG PROPERTIES
+      IMPORTED_LOCATION_DEBUG $ENV{NINTENDO_SDK_ROOT}/Libraries/${NN_SDK_PLATFORM_FOLDER}/Debug/libnn_diagTextPublicEn.lib
+      IMPORTED_LOCATION_RELEASE $ENV{NINTENDO_SDK_ROOT}/Libraries/${NN_SDK_PLATFORM_FOLDER}/Release/libnn_diagTextPublicEn.lib
+    )
+
+    add_library(NVN_LOADER STATIC IMPORTED GLOBAL)
```
Comment (pandamicro) [inline CMakeLists.txt:2126]: `NINTENDO_SDK_LOADER`

File: cmake/predefine.cmake [hunk 7]
```diff
@@ -1,12 +1,21 @@
...
-set(CC_PLATFORM_WINDOWS 2)
-set(CC_PLATFORM_ANDROID 3)
-set(CC_PLATFORM_MAC_OSX 4)
-set(CC_PLATFORM_OHOS    5)
+set(CC_PLATFORM_MAC_IOS  1)
+set(CC_PLATFORM_WINDOWS  2)
+set(CC_PLATFORM_ANDROID  3)
+set(CC_PLATFORM_MAC_OSX  4)
+set(CC_PLATFORM_OHOS     5)
+set(CC_PLATFORM_NINTENDO 6)
```
Comment (pandamicro) [inline cmake/predefine.cmake:7]: CC_PLATFORM_NX

File: cmake/predefine.cmake [hunk 8]
```diff
@@ -1,12 +1,21 @@
...
 set(CC_PLATFORM 1)
 
-if(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
+if(DEFINED ENV{NINTENDO_SDK_ROOT} AND IS_DIRECTORY ${CMAKE_CURRENT_LIST_DIR}/../cocos/renderer/gfx-nvn)
+    set(NINTENDO TRUE)
```
Comment (pandamicro) [inline cmake/predefine.cmake:11]: NX

**Review by pandamicro — CHANGES_REQUESTED** [review]

**Comment by pandamicro** [comment]
All in all, only use `NINTENDO_SDK` or `NX` as naming parts

File: CMakeLists.txt (hunk 5 above)
Comment (YunHsiao) [reply CMakeLists.txt:2093]: I think it's an internally agreed-upon acronym for `nintendo`. It's been used extensively across the SDK, i.e. libnnSdk.lib

File: CMakeLists.txt (hunk 3 above)
Comment (YunHsiao) [reply CMakeLists.txt:33]: Here we need the name of the graphics backend of NX, not NX itselff. I don't think `USE_NX` makes any sense here

File: CMakeLists.txt (hunk 6 above)
Comment (YunHsiao) [reply CMakeLists.txt:2126]: Here too. I think the name `NVN` is already (sort of) in the public domain...? Can we get at least this much leeway?

File: CMakeLists.txt (hunk 4 above)
Comment (YunHsiao) [reply CMakeLists.txt:1158]: Here too

File: CMakeLists.txt (hunk 6 above)
Comment (pandamicro) [reply CMakeLists.txt:2102]: It's simply breaking the agreement if judged. If I see the risk, I can't ignore it

File: cocos/base/threading/ThreadSafeLinearAllocator.cpp [hunk 9]
```diff
@@ -45,8 +45,8 @@ void *ThreadSafeLinearAllocator::doAllocate(size_t size, size_t alignment) noexc
     }
 
     void *   allocatedMemory = nullptr;
-    uint32_t oldUsedSize     = 0;
-    uint64_t newUsedSize     = 0; // using 64-bit integer to correctly detect overflows
+    size_t   oldUsedSize     = 0;
+    uint64_t newUsedSize     = 0; // force 64-bit here to correctly detect overflows
```
Comment (minggo) [inline cocos/base/threading/ThreadSafeLinearAllocator.cpp:49]: 这两个类型最好统一一下吧。引擎现在一般不使用 size_t

Comment (YunHsiao) [reply cocos/base/threading/ThreadSafeLinearAllocator.cpp:49]: vector 之类的返回值就是这个啊，该用就用呗

Comment (minggo) [reply cocos/base/threading/ThreadSafeLinearAllocator.cpp:49]: 这得有清晰的规则，否则引擎其它地方就不知道怎么用了。而且：
- 该函数连 oldUsedSize, newUsedSize 不统一
- 其它代码使用的话都需要类型转换

**Pull request merged**